### ISSUE TITLE
Improve CLI prompts for username and password

### ISSUE THREAD

**Issue by krzychukula** [opctl/opctl#745]
What I have to do:
```
➜  opctl run something 
-
  Please provide "password".
  Description: password used to auth w/ the pkg source
-
-
  Please provide "username".
  Description: username used to auth w/ the pkg source
-
krzychukula
```
What is wrong:
1. Username prompt should be first.
2. Username prompt HAVE to say what kind of username it wants. In this case it wants Github username but I have to guess.
3. I use 2FA so I can't really use Github password. It should be handled or:
4. It could ask for Github Auth Token

**Comment by chrisdostert** [comment]
@krzychukula thanks for filing!
re: password: github guides to use accessToken as password in their 2FA docs; I know that's slightly confusing but kinda makes sense because they're limited to auth modes supported by git. 
see [their 2fa docs](https://help.github.com/en/github/authenticating-to-github/accessing-github-using-two-factor-authentication#using-two-factor-authentication-with-the-command-line)
> After you've enabled 2FA, you must create a personal access token to use as a password when authenticating to GitHub on the command line using HTTPS URLs.

re: switching ordering of username and password prompts and showing who the auth prompt is triggered by; totally. Those seem like great UX enhancements.

**Comment by krzychukula** [comment]
What do you think about adding the info about access token to the password prompt?

Btw, speaking about git. I already have it set up. Why do I need to create yet another token? Can’t opctl reuse it?

**Comment by chrisdostert** [comment]
Opctl uses git http protocol to pull ops from arbitrary sources (similar to what go does for packages). We could add more intelligence/vendor specific messaging by seeing the source happens to be github. I agree it would be a nice enhancement.

re: git already being setup on your host) opctl doesn't actually use the git binary or config from your machine. 

The overarching goal around that was:
1) so you don't have to have git installed to run ops
1) encapsulation; ops are designed to only have access to their inputs; this keeps them composable (think react components w/ external state being passed through props).

So that's the original intent; now let the thinking ensue  : ) 

Also this could relate to https://github.com/opctl/opctl/issues/744 depending on discussion there.

**Comment by krzychukula** [comment]
Thank you for looking into it.

Let me rephrase it. 
How do I know what opctl is asking me for?
I have no idea what it can be:
1. My machine user/pass (`sudo`?)
2. Current repo Github credentials?
3. My company identity provider credentials?
4. Something else? Maybe tries do access some of my private dependencies?

> We could add more intelligence/vendor specific messaging by seeing the source happens to be github. 

I strongly disagree with `happens to be`. Git doesn't have credentials. I can do `git init` on my machine without any security. Github, Gitlab, Bitbucket, and others use credentials. I need to know what `opctl` is asking of me. In this case, I need to know it's asking about Github.

> re: git already being setup on your host) opctl doesn't actually use the git binary or config from your machine at present.

I understand the benefits here. Fewer dependencies make it easier for the user. Not dealing with old/weird git installed by the user makes opctl implementation simpler. I just want to say that the fact that it doesn't require git to be installed doesn't mean that it can't use installed git if it's available. Or at least try (because depending on it is a whole new level of problems). I agree that it's a tradeoff between making it even easier to use vs complicating the implementation. 

> so you don't have to have git installed to run ops

Hm... That's interesting. What's the use case? Bootstrapping a new project or something else?

> encapsulation; ops are designed to only have access to their inputs; this keeps them composable (think react components w/ external state being passed through props).

If I try to compose two ops will I be asked for my credentials twice?

**Comment by chrisdostert** [comment]
I think maybe small miscommunication; I was intending to say I totally agree with the proposal to include the ref to the op requiring auth, or something indicating what needs auth. I think it's a nice enhancement. 

> Hm... That's interesting. What's the use case? Bootstrapping a new project or something else?

It's just one fewer dependency. The only dependency at present is Docker and that will go away at some point. An example where it comes in handy is if you're running on a CI server or just any environment you're running opctl stand alone. 

> If I try to compose two ops will I be asked for my credentials twice?

This prompt is only for the root op; You're not in an op yet. Once you're in an op the runtime is virtual, everything must be an explicitly declared input or output of each op; nothing else is accessible from your host. In ops, you specify [op pullCreds](https://opctl.io/docs/reference/opspec/op-directory/op/call/op#pullcreds) or [container image pullCreds](https://opctl.io/docs/reference/opspec/op-directory/op/call/container/image#pullcreds) to allow pulling them from private sources.
